### PR TITLE
Add timeout to Provide call

### PR DESCRIPTION
**Pull request by michaelavila**
This just adds the the ability to set a timeout that is used by [go-bitswap](https://github.com/ipfs/go-bitswap/blob/master/workers.go#L126) when doing a `Provide`. Partially addresses https://github.com/ipfs/go-ipfs-provider/issues/7.

This also makes the worker count configurable.

### CODE REVIEW

**Comment by michaelavila** [comment]
@postables, @hsanjuan do you mind taking a look and reviewing this PR?

File: simple/provider.go [hunk 1]
```diff
@@ -14,7 +15,10 @@ import (
 
 var logP = logging.Logger("provider.simple")
 
-const provideOutgoingWorkerLimit = 8
+const (
+	provideOutgoingWorkerLimit = 8
```
Comment (hsanjuan) [inline simple/provider.go]: I think this doesn't belong here?

**Review by hsanjuan — CHANGES_REQUESTED** [review]
LGTM but I would prefer this was configurable. 3 minute is a bit arbitrary (being that it's a timeout for a content routing implementation and we don't know what `Provide` involves in it).

I also don't know if setting a timeout here might have unwanted side-effects. If it's just working around a DHT bug, what happens when that is fixed? Will 3 minutes still be the ok value? Or will disabling the timeout altogether be the right approach because the DHT will timeout by itself? What I mean is that it should probably default to "no timeout" (current behaviour) and optionally be configurable to timeout.

In short, probably exporting `ProvideTimeout` as a variable with a 0 default meaning "no timeout" is enough.

File: simple/provider.go (hunk 1 above)
Comment (michaelavila) [reply simple/provider.go]: This does belong here, but I went ahead and made it configurable.

Comment (hsanjuan) [reply simple/provider.go]: Sorry, I think I only saw I subdiff and didn't see the whole thing before.

**Comment by michaelavila** [comment]
@hsanjuan that's all fair and good feedback, thanks. I took that into consideration and made some changes. Let me know what you think.

File: simple/provider.go [hunk 2]
```diff
@@ -53,15 +81,17 @@ func (p *Provider) Provide(root cid.Cid) error {
...
 func (p *Provider) handleAnnouncements() {
-	for workers := 0; workers < provideOutgoingWorkerLimit; workers++ {
+	for workers := 0; workers < p.workerLimit; workers++ {
 		go func() {
 			for p.ctx.Err() == nil {
 				select {
 				case <-p.ctx.Done():
 					return
 				case c := <-p.queue.Dequeue():
+					ctx, cancel := context.WithTimeout(p.ctx, p.timeout)
```
Comment (hsanjuan) [inline simple/provider.go]: Right now default timeout is 0 and the context will be cancelled immediately.

**Review by hsanjuan — CHANGES_REQUESTED** [review]

File: simple/provider.go (hunk 2 above)
Comment (hsanjuan) [reply simple/provider.go]: I suggest changing the `mockRouting` in the tests to fail if the context is expired.

Comment (michaelavila) [reply simple/provider.go]: Yeah, I'll add a few tests around timing out.

**Comment by Stebalien** [comment]
I believe this will simply cause every single provide to timeout. Unfortunately, a timed-out provide is almost always completely useless (i.e., doesn't put anything in the DHT).

**Comment by bonedaddy** [comment]
I have done some testing of this PR, the revision I'm using according to `go.mod` is `v0.0.0-20190612202929-5f6a572aacdc`

Findings are recorded in the corresponding github [issue](https://github.com/ipfs/go-ipfs-provider/issues/7#issuecomment-501509183)

**Comment by michaelavila** [comment]
@Stebalien I think you're right and we just have some unexpected behavior, though I also think the issue has been around for a long time, which is why the timeout exists for the same call in `go-bitswap`. My thought here is to apply this fix as a band-aid, simply because it works and folks are being impacted, while we spend some time trying to find the underlying problem. I worry that we are not going to find a solution quickly. Thoughts?

**Comment by Stebalien** [comment]
Discussed in the standup, apparently we _ignore_ the context error in `GetClosestPeers` because we get it after we return a peer channel. Additionally, the underlying query will return _both_ the closest peers _and_ an error.

That means we'll be able to send provider records as long as none of these requests block for too long (and check the context).

As discussed, we should add a separate timeout (internally) on `GetClosestPeers` to work around this (for now).

File: simple/provider.go [hunk 3]
```diff
@@ -53,15 +81,24 @@ func (p *Provider) Provide(root cid.Cid) error {
...
 			for p.ctx.Err() == nil {
 				select {
 				case <-p.ctx.Done():
 					return
 				case c := <-p.queue.Dequeue():
+					var ctx context.Context
+					var cancel context.CancelFunc
+					if p.timeout > 0 {
+						ctx, cancel = context.WithTimeout(p.ctx, p.timeout)
+						defer cancel()
```
Comment (Stebalien) [inline simple/provider.go]: Defer in a loop.

**Review by Stebalien — CHANGES_REQUESTED** [review]
We now have https://github.com/libp2p/go-libp2p-kad-dht/pull/351 so LGTM modulo the defer issue.

**Review by Stebalien — APPROVED** [review]

**Comment by Stebalien** [comment]
I believe all the issues have now been fixed.
